pkg/server/endpoints/node/handler.go

```diff
@@ -480,7 +488,70 @@ func (h *Handler) FetchJWTSVID(ctx context.Context, req *node.FetchJWTSVIDReques
 }
 
 func (h *Handler) PushJWTKeyUpstream(ctx context.Context, req *node.PushJWTKeyUpstreamRequest) (resp *node.PushJWTKeyUpstreamResponse, err error) {
-	return nil, status.Error(codes.Unimplemented, "cannot push JWK upstream")
+	counter := telemetry_server.StartNodeAPIPushJWTKeyUpstreamCall(h.c.Metrics)
+	defer counter.Done(&err)
+	log := h.c.Log.WithField(telemetry.Method, telemetry.PushJWTKeyUpstream)
+
+	_, ok := getPeerCertificate(ctx)
+	if !ok {
+		log.Error("Downstream SVID is required for this request")
+		return nil, status.Error(codes.InvalidArgument, "downstream SVID is required for this request")
+	}
+
+	_, ok = getDownstreamEntry(ctx)
+	if !ok {
+		log.Error("Downstream entry is required for this request")
+		return nil, status.Error(codes.InvalidArgument, "downstream entry is required for this request")
+	}
+
+	err = h.limiter.Limit(ctx, PushJWTKey, 1)
+	if err != nil {
+		log.WithError(err).Error("Rejecting request due to JWK push rate limiting")
+		return nil, status.Error(codes.ResourceExhausted, err.Error())
+	}
+
+	upstreamAuth, useUpstream := h.c.Catalog.GetUpstreamAuthority()
+	if useUpstream { // We are an intermediate SPIRE server
+		publishCtx, cancel := context.WithTimeout(ctx, h.c.PublishJWKTimeout)
+		defer cancel()
+		res, err := upstreamAuth.PublishJWTKey(
+			publishCtx,
+			&upstreamauthority.PublishJWTKeyRequest{
+				JwtKey: req.JwtKey,
+			})
+
+		switch {
+		case gcu.IsUnimplementedError(err):
+			h.warnOnce.Do(func() {
+				log.Warn("UpstreamAuthority does not support JWT-SVIDs. Workloads managed " +
+					"by this server may have trouble communicating with workloads outside " +
+					"this cluster when using JWT-SVIDs.")
+			})
+		case err != nil:
+			return nil, status.Error(codes.Internal, err.Error())
+		default:
+			// Send back to the caller the received JWKs from upstream,
+			// this will change soon according to https://github.com/spiffe/spire/issues/1372#issuecomment-589469449
+			return &node.PushJWTKeyUpstreamResponse{
+				JwtSigningKeys: res.UpstreamJwtKeys,
+			}, nil
+		}
+	}
+
+	if err := h.c.Manager.AppendBundle(ctx, nil, []*common.PublicKey{req.JwtKey}); err != nil {
+		return nil, err
+	}
+
+	// Ensure we delete any cached bundle to get the latest one.
+	h.dsCache.DeleteBundleEntry(h.c.TrustDomain.String())
+	bundle, err := h.getBundle(ctx, h.c.TrustDomain.String())
```

Comment:
will it try to remove x509 bundle too?

Reply:
The datastore cache caches datastore.FetchBundleResponse objetcs, so if you are referring to the Root CA certificates list that are included in common.Bundle, then yes, it will.

Reply:
can it affect us on somthing? basically we are updating RootCAs when calling an endpoint

Reply:
I don't believe so, it just invalidates an entry in a cache, so next time a caller uses dsCache the entry, it will be created from fresh data.